File: backend/src/main/java/cn/etarch/mao/audit/interceptor/AuditInterceptor.java
package cn.etarch.mao.audit.interceptor;

import cn.etarch.mao.audit.entity.AuditLog;
import cn.etarch.mao.audit.service.AuditLogService;
import cn.etarch.mao.user.entity.User;
import cn.etarch.mao.user.mapper.UserMapper;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.HandlerInterceptor;

import java.util.Set;

@Component
@RequiredArgsConstructor
public class AuditInterceptor implements HandlerInterceptor {

    private static final int MAX_QUERY_LENGTH = 1024;
    private static final int MAX_ERROR_LENGTH = 1024;
    private static final Set<String> AUDITED_PREFIXES = Set.of(
            "/v1/agents",
            "/v1/models",
            "/v1/users",
            "/v1/roles",
            "/v1/permissions",
            "/v1/skill-docs",
            "/v1/admin",
            "/v1/system-settings",
            "/v1/notifications"
    );

    private final AuditLogService auditLogService;
    private final UserMapper userMapper;

    @Override
    public void afterCompletion(HttpServletRequest request, HttpServletResponse response, Object handler, Exception ex) {
        String path = request.getRequestURI();
        String contextPath = request.getContextPath();
        if (contextPath != null && !contextPath.isBlank() && path.startsWith(contextPath)) {
            path = path.substring(contextPath.length());
        }
        if (!shouldAudit(path)) {
            return;
        }

        AuditLog log = new AuditLog();
        log.setMethod(request.getMethod());
        log.setPath(path);
        log.setQueryString(truncate(request.getQueryString(), MAX_QUERY_LENGTH));
        log.setIp(resolveIp(request));
        log.setStatus(response.getStatus());
        log.setSuccess(ex == null && response.getStatus() < 400 ? 1 : 0);
        log.setErrorMessage(ex != null ? truncate(ex.getMessage(), MAX_ERROR_LENGTH) : null);
        log.setAction(resolveAction(request.getMethod()));
        log.setObjectType(resolveObjectType(path));
        log.setObjectId(resolveObjectId(path));

        Object principal = SecurityContextHolder.getContext().getAuthentication() != null
                ? SecurityContextHolder.getContext().getAuthentication().getPrincipal()
                : null;
        if (principal instanceof Long userId) {
            log.setUserId(userId);
            User user = userMapper.selectById(userId);
            if (user != null) {
                log.setUsername(user.getUsername());
            }
        }

        try {
            auditLogService.record(log);
        } catch (Exception ignored) {
            // Audit must never break the business request.
        }
    }

    private boolean shouldAudit(String path) {
        if (path == null || path.startsWith("/v1/auth") || path.startsWith("/v1/audit/logs")) {
            return false;
        }
        return AUDITED_PREFIXES.stream().anyMatch(path::startsWith);
    }

    private String resolveAction(String method) {
        return switch (method) {
            case "POST" -> "CREATE";
            case "PUT", "PATCH" -> "UPDATE";
            case "DELETE" -> "DELETE";
            case "GET" -> "READ";
            default -> "EXECUTE";
        };
    }

    private String resolveObjectType(String path) {
        if (path == null) return "unknown";
        String[] parts = path.split("/");
        if (parts.length >= 3) {
            if ("admin".equals(parts[2]) && parts.length >= 4) {
                return "admin." + parts[3];
            }
            return parts[2];
        }
        return "unknown";
    }

    private String resolveObjectId(String path) {
        if (path == null) return null;
        String[] parts = path.split("/");
        for (int i = 3; i < parts.length; i++) {
            if (parts[i].matches("\\d+")) {
                return parts[i];
            }
        }
        return null;
    }

    private String resolveIp(HttpServletRequest request) {
        String forwarded = request.getHeader("X-Forwarded-For");
        if (forwarded != null && !forwarded.isBlank()) {
            return forwarded.split(",")[0].trim();
        }
        String realIp = request.getHeader("X-Real-IP");
        return realIp != null && !realIp.isBlank() ? realIp : request.getRemoteAddr();
    }

    private String truncate(String value, int max) {
        if (value == null || value.length() <= max) {
            return value;
        }
        return value.substring(0, max);
    }
}

File: backend-ts/src/audit/audit.interceptor.ts
import type { AuditLogService } from './audit.service.js';
import type { AuditLog, AuditRecordInput } from './types.js';

const MAX_QUERY_LENGTH = 1024;
const MAX_ERROR_LENGTH = 1024;
const AUDITED_PREFIXES = [
  '/v1/agents',
  '/v1/models',
  '/v1/users',
  '/v1/roles',
  '/v1/permissions',
  '/v1/skill-docs',
  '/v1/admin',
  '/v1/system-settings',
];
/**
 * 只审计变更方法的前缀：这些资源的读接口仍对 session:read 开放，
 * 但跨用户改他人数据需要留痕（定时任务是代表用户长期自动执行的任务，改动影响面大）。
 */
const AUDITED_WRITE_PREFIXES = ['/v1/scheduled-tasks'];
/** 读取型 POST（每次编辑击键都会调用的预览接口），审计它们只会淹没日志。 */
const AUDITED_WRITE_EXCLUDES = ['/v1/scheduled-tasks/cron-preview'];
const WRITE_METHODS = new Set(['POST', 'PUT', 'PATCH', 'DELETE']);

export function shouldAudit(path: string | null | undefined, method?: string): boolean {
  if (path == null || path.startsWith('/v1/auth') || path.startsWith('/v1/audit/logs')) {
    return false;
  }
  if (AUDITED_PREFIXES.some((prefix) => path.startsWith(prefix))) {
    return true;
  }
  if (!WRITE_METHODS.has(method ?? '')) {
    return false;
  }
  if (AUDITED_WRITE_EXCLUDES.some((prefix) => path.startsWith(prefix))) {
    return false;
  }
  return AUDITED_WRITE_PREFIXES.some((prefix) => path.startsWith(prefix));
}

export function resolveAction(method: string): string {
  switch (method) {
    case 'POST':
      return 'CREATE';
    case 'PUT':
    case 'PATCH':
      return 'UPDATE';
    case 'DELETE':
      return 'DELETE';
    case 'GET':
      return 'READ';
    default:
      return 'EXECUTE';
  }
}

export function resolveObjectType(path: string | null | undefined): string {
  if (path == null) {
    return 'unknown';
  }
  const parts = path.split('/');
  if (parts.length >= 3) {
    if (parts[2] === 'admin' && parts.length >= 4) {
      return `admin.${parts[3]}`;
    }
    return parts[2];
  }
  return 'unknown';
}

export function resolveObjectId(path: string | null | undefined): string | null {
  if (path == null) {
    return null;
  }
  const parts = path.split('/');
  for (let i = 3; i < parts.length; i++) {
    if (/^\d+$/.test(parts[i])) {
      return parts[i];
    }
  }
  return null;
}

export function resolveIp(headers: Record<string, string | string[] | undefined>, remoteAddr?: string | null): string | undefined {
  const forwarded = headerValue(headers, 'x-forwarded-for');
  if (forwarded && forwarded.trim()) {
    return forwarded.split(',')[0].trim();
  }
  const realIp = headerValue(headers, 'x-real-ip');
  return realIp && realIp.trim() ? realIp : (remoteAddr ?? undefined);
}

export async function recordAudit(service: AuditLogService, input: AuditRecordInput): Promise<void> {
  if (!shouldAudit(input.path, input.method)) {
    return;
  }
  const log: AuditLog = {
    method: input.method,
    path: input.path,
    queryString: truncate(input.queryString, MAX_QUERY_LENGTH),
    ip: input.ip,
    status: input.status,
    success: input.errorMessage == null && input.status < 400 ? 1 : 0,
    errorMessage: input.errorMessage != null ? truncate(input.errorMessage, MAX_ERROR_LENGTH) : null,
    action: resolveAction(input.method),
    objectType: resolveObjectType(input.path),
    objectId: resolveObjectId(input.path),
    userId: input.userId,
    username: input.username,
  };
  try {
    await service.record(log);
  } catch {
    // Audit must never break the business request.
  }
}

function truncate(value: string | null | undefined, max: number): string | null | undefined {
  if (value == null || value.length <= max) {
    return value;
  }
  return value.slice(0, max);
}

/** 供拦截器之外的审计写入方（如云端终端）复用同一截断策略。 */
export function truncateAuditError(value: string | null | undefined): string | null {
  return truncate(value, MAX_ERROR_LENGTH) ?? null;
}

function headerValue(
  headers: Record<string, string | string[] | undefined>,
  name: string,
): string | undefined {
  const raw = headers[name] ?? headers[name.toLowerCase()];
  if (Array.isArray(raw)) {
    return raw[0];
  }
  return raw;
}
